File: example/go-redis/redis/hashes.go
package redis

import (
	"context"
	"errors"
	"fmt"

	redisv9 "github.com/redis/go-redis/v9"
)

func (r *repository) HSET(ctx context.Context, key string, values map[string]interface{}) error {
	return r.client.HSet(ctx, key, values).Err()
}

func (r *repository) HGET(ctx context.Context, key, field string) (string, error) {
	result, err := r.client.HGet(ctx, key, field).Result()
	if err != nil {
		if errors.Is(err, redisv9.Nil) {
			return "", fmt.Errorf("field not found")
		}

		return "", err
	}

	return result, nil
}

func (r *repository) HGETALL(ctx context.Context, key string) (map[string]string, error) {
	return r.client.HGetAll(ctx, key).Result()
}

File: example/rust-redis/src/redis/hashes.rs
use std::collections::HashMap;

use anyhow::{anyhow, Result};
use redis::AsyncCommands;

use super::redis::Repository;

impl Repository {
    pub async fn hset(&self,key: &str,values: &HashMap<String, String>) -> Result<()> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;

        conn.hset_multiple::<_, _, _, ()>(
            key,
            &values
                .iter()
                .map(|(k, v)| (k.as_str(), v.as_str()))
                .collect::<Vec<_>>(),
        )
        .await?;

        Ok(())
    }

    pub async fn hget(&self,key: &str, field: &str) -> Result<String> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;
        let result: Option<String> = conn.hget(key, field).await?;
        match result {
            Some(value) => Ok(value),
            None => Err(anyhow!("field not found")),
        }
    }

    pub async fn hgetall(&self,key: &str) -> Result<HashMap<String, String>> {
        let mut conn = self.client.get_multiplexed_async_connection().await?;
        let result: HashMap<String, String> =
            conn.hgetall(key).await?;
        Ok(result)
    }
}

